File: wsc/public/js/student_admission.js
frappe.ui.form.on('Student Admission', {
    before_save(frm){
        if(frm.doc.admission_start_date && frm.doc.admission_end_date && frm.doc.admission_end_date < frm.doc.admission_start_date){
            
            frappe.throw("Counselling End date should be greater than Counselling Start date");
        
        }
        if(frm.doc.enrollment_end_date && frm.doc.enrollment_start_date && frm.doc.enrollment_end_date < frm.doc.enrollment_start_date){
            
            frappe.throw("Enrollment End date should be greater than Enrollment Start date");
        
        }
    },
    setup(frm) {
        frm.set_df_property('program_grade', 'reqd', 1);
        frm.set_query("academic_calendar", function() {
            return {
                filters: {
                    "programs":frm.doc.admission_program,
                    "program":frm.doc.semester
                }
            };
        });
        frm.set_query("counselling_structure", function() {
            return {
                query: 'wsc.wsc.doctype.student_admission.get_counselling_structure',
                filters: {
                    "program_grade":frm.doc.program_grade,
                    "programs":frm.doc.admission_program,
                    "academic_year":frm.doc.academic_year
                }
            };
        });
        frm.set_query("admission_program", function() {
            return {
                filters: {
                    "program_grade":frm.doc.program_grade
                }
            };
        });
        frm.set_query("fee_structure","hostel_fees_", function(){
            return{
                filters:{
                    "programs":frm.doc.admission_program,
                    "docstatus":1
                }
            }
        })
        frm.set_query("fee_structure","counselling_fee_structures", function(){
            return{
                filters:{
                    "fee_type":"Counselling Fees",
                    "programs":frm.doc.admission_program ,
                    "docstatus":1
                }
            }
        })
        frm.set_query("fee_structure","admission_fee_structure", function(){
            return{
                filters:{
                    "fee_type":"Admission Fees",
                    "programs":frm.doc.admission_program,
                    "docstatus":1
                }
            }
        })
		frm.set_query("semester", function() {
			return {
				query: 'wsc.wsc.doctype.student_admission.get_sem',
				filters: {
					"program":frm.doc.admission_program
				}
			};
		});
        frm.set_query("academic_term", function() {
            return {
                filters: {
                    "academic_year":frm.doc.academic_year
                }
            };
        });
        frm.set_query("document_type","reservations_distribution", function(){
            return{
                query: 'wsc.wsc.doctype.student_admission.get_doc'
            }
        })
        frm.fields_dict['reservations_distribution'].grid.get_field('seat_reservation_type').get_query = function(doc){
            let seat_reservation_type_list = [];
            if(!doc.__islocal) seat_reservation_type_list.push(doc.seat_reservation_type);
            $.each(doc.reservations_distribution, function(idx, val){
                if (val.seat_reservation_type) seat_reservation_type_list.push(val.seat_reservation_type);
            });
            return { filters: [['Seat Reservation Type', 'name', 'not in', seat_reservation_type_list]] };
        };
        // frm.fields_dict['required_documents_list'].grid.get_field('student_category').get_query = function(doc){
        //     let document_type_list = [];

        //     // if(!doc.__islocal) document_type_list.push(doc.document_type, doc.student_category);
        //     console.log("000000000document_type_list ",document_type_list)
        //     $.each(doc.required_documents_list, function(idx, val){

        //         if (val.document_type) document_type_list.push({'document_type':val.document_type});
        //         if (val.student_category) document_type_list.push({'student_category':val.student_category});
        //     });
        //     console.log("1111111document_type_list ",document_type_list)
           
        //     return { filters: [['Documents Template', 'name', 'in', document_type_list],['Student Category', 'name', 'not in', document_type_list],['Documents Template', "is_active","=",1]] };
        // };
         frm.set_query("document_type","required_documents_list", function() {
            return {
                filters: {
                    "is_active":1
                }
            };
        });

    }
})


frappe.ui.form.on("Reservations List", "update_balance", function(frm, cdt, cdn) {
        var d = locals[cdt][cdn];
        let dialog = new frappe.ui.Dialog({
        title: __('Update Seats'),
        fields: [
            {
                "label" : "Type",
                "fieldname": "type",
                "fieldtype": "Select",
                "options":"\nAdd Balance\nDeduct Balance",
                "reqd":1
            },
            {
                "label" : "No of Seats",
                "fieldname": "no_of_seats",
                "fieldtype": "Int"
            }
        ],
        primary_action: function() {
            var data=dialog.get_values();
            if (data.type=="Add Balance"){
                frappe.model.set_value(cdt, cdn, "allocated_seat", (d.allocated_seat || 0)+data.no_of_seats);
                frappe.model.set_value(cdt, cdn, "seat_balance", (d.seat_balance || 0)+data.no_of_seats);
            }
            else{
                frappe.model.set_value(cdt, cdn, "allocated_seat", (d.allocated_seat || 0)-data.no_of_seats);
                frappe.model.set_value(cdt, cdn, "seat_balance", (d.seat_balance || 0)-data.no_of_seats);
            }
            dialog.hide();
        },
        primary_action_label: __('Update')
        });
        dialog.show();
       
});
frappe.ui.form.on("Eligibility Parameter List", "percentagecgpa", function(frm, cdt, cdn) {
    frappe.model.set_value(cdt, cdn, "total_score", 0);
    frappe.model.set_value(cdt, cdn, "eligible_score", 0);
});
frappe.ui.form.on("Required Documents", "document_type", function(frm, cdt, cdn) {
    var d = locals[cdt][cdn];
    frappe.db.get_value("Documents Template", {'name':d.document_type},'student_category', resp => {
        frappe.model.set_value(cdt, cdn, "student_category", resp.student_category);
    })
});

File: wsc/wsc/validations/student_admission.py
import frappe
import collections
from collections import Counter
from wsc.wsc.utils import semester_belongs_to_programs,academic_term


def validate(doc, method):
    validate_program_grade(doc)
    # validate_academic_year(doc)
    validate_semester(doc)
    validate_fee_structure_in_hostel_fees(doc)
    validate_fee_structure_in_counselling(doc)
    validate_fee_structure_in_admission(doc)
    validate_academic_calender(doc)
    validate_counselling_structure(doc)
    validate_admission_program(doc)
    validate_total_seats(doc)
    validate_eligibility_parameters(doc)
    validate_document_type(doc)
    academic_term(doc)
    validate_total_seats(doc)
    validate_parameters(doc)
    semester_belongs_to_programs(doc)

def validate_total_seats(doc):
    total_seats = 0
    if len(doc.reservations_distribution) > 0:
        a = [s.seat_reservation_type for s in doc.reservations_distribution]
        seat_reservation_type_list = [d for d, count in collections.Counter(a).items() if count > 1]
        if seat_reservation_type_list: 
            frappe.throw("Duplicate Seat Reservations type in Reservations Distributions.")
    for rd in doc.reservations_distribution:
        if rd.allocated_seat:
            total_seats += rd.allocated_seat
            if rd.seat_balance > rd.allocated_seat:
                frappe.throw("Seats balance should be less than or equal to allocated seats for")
    if doc.total_seats < total_seats and doc.total_seats > 0:
        frappe.throw("Totals of allocated seats should be less than or equal to total seats")
    elif doc.total_seats < 0:
        frappe.throw("Total Seats should not be less than zero")

def validate_eligibility_parameters(doc):
    #duplicate row validation
    if len(doc.eligibility_parameter_list) > 0:
        dataList = []
        for e in doc.eligibility_parameter_list:
            row = {}
            row.update({'student_category':e.student_category, 'parameter':e.parameter})
            dataList.append(row)
        res_list = [i for n, i in enumerate(dataList) if i not in dataList[n + 1:]]
        if len(res_list) < len(dataList):
            frappe.throw("Duplicate row in Eligibility Parameter List.")

    for eligibility in doc.get("eligibility_parameter_list"):
        # negative value
        if eligibility.total_score < 0 or eligibility.eligible_score < 0:
            frappe.throw("Score Value Must be <b>Positive</b> In  {0}".format(doc.doctype))

        # total value should be greater
        if eligibility.total_score < eligibility.eligible_score:
            frappe.throw("<b>Eligible Score</b>  Should be Less than or Equal to <b>Total Score</b> In {0}".format(doc.doctype))


def validate_admission_program(doc):
    if doc.counselling_structure and doc.admission_program:
        if doc.admission_program not in [d.programs for d in frappe.get_all("Counselling Programs", {"parent":doc.counselling_structure},['programs'])]:
            frappe.throw("Admission program <b>'{0}'</b> not belongs to counselling structure <b>'{1}'</b> ".format(doc.admission_program, doc.counselling_structure))

def validate_counselling_structure(doc):
    if doc.counselling_structure:
        if doc.counselling_structure not in [d.name for d in frappe.get_all("Counselling Structure", {"program_grade":doc.program_grade},['name'])]:
            frappe.throw("Counselling Structure <b>'{0}'</b> not belongs to program grade <b>'{1}'</b> ".format(doc.counselling_structure, doc.program_grade))

def validate_academic_calender(doc):
    if doc.academic_calendar:
        if doc.academic_calendar not in [d.name for d in frappe.get_all("Academic Calendar Template", { "programs":doc.admission_program,
                        "program":doc.semester},['name'])]:
            frappe.throw("Academic Calendar <b>'{0}'</b> not belongs to program <b>'{1}'</b> and semester <b>'{2}'</b>".format(doc.get('academic_calendar'), doc.get('admission_program'),doc.get('semester')))

def validate_program_grade(doc):
    if doc.admission_program not in [d.name for d in frappe.get_all("Programs", {'program_grade':doc.get('program_grade')},['name'])]:
        frappe.throw("Admission Program <b>'{0}'</b> not belongs to program grade <b>'{1}'</b>".format( doc.get('admission_program'),doc.get('program_grade')))

def validate_semester(doc):
    if doc.semester not in [d.semesters for d in frappe.get_all("Semesters", {'parent':doc.get('admission_program')},['semesters'])]:
        frappe.throw("Semester <b>'{0}'</b> not belongs to program <b>'{1}'</b>".format(doc.get('semester'), doc.get('admission_program')))

def validate_academic_year(doc):
    if doc.academic_term:
        if doc.academic_term not in [d.name for d in frappe.get_all("Academic Term", {'academic_year':doc.get('academic_year')},['name'])]:
            frappe.throw("Academic Term <b>'{0}'</b> not belongs to academic year <b>'{1}'</b>".format(doc.get('academic_term'), doc.get('academic_year')))

def validate_fee_structure_in_hostel_fees(doc):
    for i in doc.hostel_fees_:
        if i.fee_structure:
            if i.fee_structure not in [d.name for d in frappe.get_all("Fee Structure", {'programs':doc.get('admission_program'),"docstatus":1},['name'])]:
                frappe.throw("Fee structure <b>'{0}'</b> not belongs to program <b>'{1}'</b> in hostel fees".format(i.fee_structure, doc.get('admission_program')))

def validate_fee_structure_in_counselling(doc):
    for i in doc.counselling_fee_structures:
        if i.fee_structure:
            if i.fee_structure not in [d.name for d in frappe.get_all("Fee Structure", {"fee_type":"Counselling Fees",'programs':doc.get('admission_program'),"docstatus":1},['name'])]:
                frappe.throw("Fee structure <b>'{0}'</b> not belongs to program <b>'{1}'</b> in counselling fee structures".format(i.fee_structure, doc.get('admission_program')))

def validate_fee_structure_in_admission(doc):
    for i in doc.admission_fee_structure:
        if i.fee_structure:
            if i.fee_structure not in [d.name for d in frappe.get_all("Fee Structure", {"fee_type":"Admission",'programs':doc.get('admission_program'),"docstatus":1},['name'])]:
                frappe.throw("Fee structure <b>'{0}'</b> not belongs to program <b>'{1}'</b> in admission fee structures".format(i.fee_structure, doc.get('admission_program')))

def validate_document_type(doc):
    if len(doc.required_documents_list) > 0:
        dataList = []
        for i in doc.required_documents_list:
            if i.document_type:
                if i.document_type not in [d.name for d in frappe.get_all("Documents Template", {"is_active":1},['name'])]:
                    frappe.throw("Document type <b>'{0}'</b> is not active.".format(i.document_type))
                #duplicate row validation
                row = {}
                row.update({'student_category':i.student_category, 'document_type':i.document_type})
                dataList.append(row)
        res_list = [i for n, i in enumerate(dataList) if i not in dataList[n + 1:]]
        if len(res_list) < len(dataList):
            frappe.throw("Duplicate row in Required Documents List.")

@frappe.whitelist()
def get_sem(pdoctype, txt, searchfield, start, page_len, filters):
    fltr={"parent":filters.get("program")}
    lst = []
    if txt:
        fltr.update({"semesters":['like', '%{}%'.format(txt)]})
    for i in frappe.get_all("Semesters",fltr,['semesters']):
        if i.semesters not in lst:
            lst.append(i.semesters)
    return [(d,) for d in lst]

@frappe.whitelist()
def get_doc(pdoctype, txt, searchfield, start, page_len, filters):
    doc_list = ['Program Grades', 'Programs', 'Program', 'Student Category', 'Academic Year','Academic Term', 'Student Admission', 'Counselling Structure', 'Bank', 'Bank Account Type', 'Student Applicant']
    return frappe.get_all("DocType" , {"name":["in", doc_list]}, ['name'],as_list = 1)

def validate_total_seats(doc):
    if doc.total_seats==0:
        frappe.throw("Please fill the <b>Total Seats</b> & Set <b>Reservations Distribution</b>")

def validate_parameters(doc):
    for d in doc.get("eligibility_parameter_list"):
        if d.total_score==0:
            frappe.throw("#Row {0} Please Give The <b>Total Score</b> More than Zero".format(d.idx))


@frappe.whitelist()
def get_counselling_structure(pdoctype, txt, searchfield, start, page_len, filters):
    fltr={}
    if filters.get("program_grade") and filters.get("programs") and filters.get("academic_year"):
        dept = frappe.db.get_value("Programs",filters.get("programs"),'department')
        fltr.update({
            "program_grade":filters.get("program_grade"),
            "department":frappe.db.get_value("Department",{'name':dept}, 'parent_department'),
            "academic_year":filters.get("academic_year"),
            'name': ['like', '%{}%'.format(txt)]
        })
        return frappe.get_all("Counselling Structure",fltr,['name'],as_list = 1)
    else:
        frappe.msgprint("Please Select <b>Program Grade</b> , <b>Admission Program</b> and <b>Academic Year</b>  first")
        return []